File: practice/app/rc4/php/rc4.php
<?php

/*
 * rc4加密算法
 * $pwd 密钥
 * $data 要加密的数据
 */

function rc4 ($pwd, $data)
{
    $key[] ="";
    $box[] ="";
    $cipher = ""; 

    $pwd_length = strlen($pwd);
    $data_length = strlen($data);

    for ($i = 0; $i < 256; $i++)
    {
        $key[$i] = ord($pwd[$i % $pwd_length]);
        $box[$i] = $i;
    }
    
    for ($j = $i = 0; $i < 256; $i++)
    {
        $j = ($j + $box[$i] + $key[$i]) % 256;
        $tmp = $box[$i];
        $box[$i] = $box[$j];
        $box[$j] = $tmp;
    }

   // echo implode($box, ' '), "\n";

    for ($a = $j = $i = 0; $i < $data_length; $i++)
    {
        $a = ($a + 1) % 256;
        $j = ($j + $box[$a]) % 256;

        $tmp = $box[$a];
        $box[$a] = $box[$j];
        $box[$j] = $tmp;

        $k = $box[(($box[$a] + $box[$j]) % 256)];
        $cipher .= chr(ord($data[$i]) ^ $k);
    }    
    return $cipher;
}

if (count($argv) > 3) {
    if ($argv[1] == "enb") {
        echo $data = base64_encode(rc4($argv[2], $argv[3])) . "\n";
    } else if ($argv[1] == "deb") {
        echo $data = rc4($argv[2], base64_decode($argv[3])) . "\n";
    }
}else{    
    $key = "12345678";
    $data = "梯子网@91waijiao";
    echo "key : " . $key . "\n";
    echo "raw : " . $data . "\n";
    echo "encrypt  :" . ($encrypt_data = rc4($key, $data)) . "\n";
    echo "encrypt64:" . base64_encode($encrypt_data) . "\n";
    echo "decrypt  :" . rc4($key, $encrypt_data) . "\n";
}

File: practice/app/rc4/java/RC4.java

public class RC4{
    private final byte[] s = new byte[256];
    private int x,y;

    public RC4(byte[] key) {
        for(int i=0;i<256;i++)
            s[i] = (byte)i;
        for(int i=0,j=0;i<256;i++) {
            j = (j + (s[i]&0xff) + (key[i%key.length]&0xff))&0xff;
            byte tmp = s[i];
            s[i] = s[j];
            s[j] = tmp;
        }
        for(int i=0;i<256;i++)
            ;//System.out.println(s[i]);
    }

    public int encrypt(byte[] in, int in_offset, byte[] out,
            int out_offset, int len) {
        int x = this.x;
        int y = this.y;
        byte[] s = this.s;
        for(int i=0;i<len;i++) {
            x = (x + 1) & 0xff;
            y = (y + (s[x]&0xff)) & 0xff;
            byte tmp = s[x];
            s[x] = s[y];
            s[y] = tmp;
            int t = ((s[x]&0xff) + (s[y]&0xff))&0xff;
            int k = s[t];
            out[out_offset+i] = (byte)((in[in_offset+i]&0xff)^k);
        }
        this.x = x;
        this.y = y;
        return 0;
    }

    public int decrypt(byte[] in, int in_offset, byte[] out,
            int out_offset, int len) {
        return encrypt(in, in_offset, out, out_offset, len);
    }

    public byte[] encrypt(byte[] in) {
        byte[] results = new byte[in.length];
        encrypt(in, 0, results, 0, in.length);
        return results;
    }

    public byte[] decrypt(byte[] in) {
        byte[] results = new byte[in.length];
        decrypt(in, 0, results, 0, in.length);
        return results;
    }

    public static void main(String[] args) {
        String data = "梯子网@91waijiao";
        String key = "12345678";
        RC4 rc4 = new RC4(key.getBytes());
        byte[] encrypt_data = rc4.encrypt(data.getBytes());
        System.out.println(new String(encrypt_data));
        
        RC4 rc5 = new RC4(key.getBytes());
        byte[] decrypt_data = rc5.decrypt(encrypt_data);
        System.out.println(new String(decrypt_data));
    }
}
